### ISSUE TITLE
pylint false positive attempted relative import beyond top-level package

### ISSUE THREAD

**Issue by pedroqueiroga** [flycheck/flycheck#1758]
This template is for **user support**.  If you are reporting an actual Flycheck bug, please use the relevant issue template instead.

## Checklist
- [x] I have searched existing issues and StackOverflow for solutions.
- [x] I have read the [Flycheck manual][].
- [x] I have read the [Troubleshooting Guide][].
- [x] I have checked my issue has not been reported yet.

## Problem description
python-pylint reports [relative-beyond-top-level] even though I can execute my flask project just fine. If I use pylint alone in the command line it works as expected (no relative-beyond-top-level error), and if I issue the same command as `C-c ! C-c python-pylint` it works as expected too.

example directory tree:
```
├── entry.py
└── pkg
    ├── bar.py
    ├── foo.py
    └── __init__.py
```

foo.py:
```python3
def hello_world():
    return 'Hello World'
```

bar.py:
```python3
from .foo import hello_world


def main():
    assert hello_world() == 'Hello World!'
    return True
```

entry.py:
```python3
from pkg.bar import main

print(main())
```

## Steps to reproduce
1. Open file 'pkg/bar.py'
2. Check current buffer (done automatically) 
3. pylint falsely reports error.

- [x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
No error.

## System configuration
```
Syntax checkers for buffer bar.py<pkg> in python-mode:

First checker to run:

  python-flake8
    - may enable:         yes
    - executable:         Found at /home/pedro/testvenv/bin/python3
    - configuration file: Found at "/home/pedro/test/setup.cfg"
    - `flake8' module:    Found at "/home/pedro/testvenv/lib/python3.7/site-packages/flake8/__init__.py"
    - next checkers:      python-pylint, python-mypy

Checkers that may run as part of the first checker's chain:

  python-pylint
    - may enable:         yes
    - executable:         Found at /home/pedro/testvenv/bin/python3
    - configuration file: Found at "/home/pedro/test/.pylintrc"
    - `pylint' module:    Found at "/home/testvenv/lib/python3.7/site-packages/pylint/__init__.py"
    - next checkers:      python-mypy

  python-mypy
    - may enable:         yes
    - predicate:          t
    - executable:         Found at /home/pedro/testvenv/bin/mypy
    - configuration file: Found at "/home/pedro/test/setup.cfg"

Checkers that could run if selected:

  python-pycompile  select
    - may enable:    yes
    - executable:    Found at /home/pedro/testvenv/bin/python3
    - next checkers: python-mypy

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20200427.12)
Emacs version:    26.1
System:           x86_64-pc-linux-gnu
Window system:    x
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes
I don't think this has to do with pylint since it works fine using either the command provided by `C-c ! C-c python-pylint RET` or simply using `$ pylint pkg/bar.py`. If I change to an absolute import: `from pkg.foo import hello_world` the error disappears, but I would prefer to keep using relative imports.

**Comment by cpitclaudel** [comment]
Thanks for the detailed report :+1:  I wonder if there's something missing here.  I used the following recipe:

```
$ cd /tmp/
$ mkdir python; cd python
$ printf "from pkg.bar import main\nprint(main())" > entry.py
$ mkdir pkg
$ touch pkg/__init__.py
$ printf "def hello_world():\n    return 'Hello World'" > pkg/foo.py
$ printf "from .foo import hello_world" > pkg/bar.py
```

When I do this and open bar.py in emacs -Q I don't get the error that you mentioned.  What did I do wrong?

**Comment by pedroqueiroga** [comment]
Hello, thank you for responding so quickly. I am unsure, I just copied your commands and was able to reproduce. The only package I have loaded is flycheck and things that were necessary to successfully load flycheck:
```elisp
(require 'package)
(setq package-load-list
      '((flycheck t)
	(dash t)
	(epl t)
	(pkg-info t)))
(package-initialize)
```

`*Messages*` buffer:
```
For information about GNU Emacs and the GNU system, type C-h C-a.
Can’t guess python-indent-offset, using defaults: 4
Mark set [2 times]
Unable to activate package ‘flycheck’.
Required package ‘dash-2.12.1’ is unavailable
You can run the command ‘eval-buffer’ with M-x ev-b RET
Unable to activate package ‘flycheck’.
Required package ‘dash-2.12.1’ is unavailable
Unable to activate package ‘flycheck’.
Required package ‘pkg-info-0.4’ is unavailable
Unable to activate package ‘pkg-info’.
Required package ‘epl-0.8’ is unavailable
Unable to activate package ‘pkg-info’.
Required package ‘epl-0.8’ is unavailable
Unable to activate package ‘flycheck’.
Required package ‘pkg-info-0.4’ is unavailable
user-error: End of history; no default available
Flycheck mode enabled in current buffer
'.foo.hello_world' imported but unused [F401]

Final newline missing [missing-final-newline]

Attempted relative import beyond top-level package [relative-beyond-top-level]

Unused hello_world imported from foo [unused-import]
'.foo.hello_world' imported but unused [F401]

Final newline missing [missing-final-newline]

Attempted relative import beyond top-level package [relative-beyond-top-level]

Unused hello_world imported from foo [unused-import]
mwheel-scroll: End of buffer
mwheel-scroll: Beginning of buffer
Mark set [2 times]
mwheel-scroll: Beginning of buffer
```

`*Flycheck error messages*` buffer
```
'.foo.hello_world' imported but unused [F401]

Final newline missing [missing-final-newline]

Attempted relative import beyond top-level package [relative-beyond-top-level]

Unused hello_world imported from foo [unused-import]
```

screenshot:
![image](https://user-images.githubusercontent.com/20212334/81291131-a394d180-903f-11ea-95de-48b3f923f801.png)

Should I provide more information?

**Comment by cpitclaudel** [comment]
Thanks a lot! I managed to reproduce it in a clean environment.
The problem is the directory from which pylint is invoked:

```
/tmp/python/pkg$ pylint --reports\=n /tmp/python/pkg/bar.py
************* Module bar
bar.py:1:0: C0102: Black listed name "bar" (blacklisted-name)
bar.py:1:0: C0114: Missing module docstring (missing-module-docstring)
bar.py:1:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
bar.py:1:0: W0611: Unused hello_world imported from foo (unused-import)

/tmp/python$ pylint --reports\=n /tmp/python/pkg/bar.py
************* Module pkg.bar
pkg/bar.py:1:0: C0102: Black listed name "bar" (blacklisted-name)
pkg/bar.py:1:0: C0114: Missing module docstring (missing-module-docstring)
pkg/bar.py:1:0: W0611: Unused hello_world imported from foo (unused-import)
```

The problem is, I have no idea what the right algorithm is to determine where to run Flycheck from.  The documentation is here: http://pylint.pycqa.org/en/latest/faq.html#can-i-give-pylint-a-file-as-an-argument-instead-of-a-module, but it doesn't help much.  It's true that `pylint pkg.bar` works, but how are we supposed to guess that this is what is desired? :/

**Comment by cpitclaudel** [comment]
Related issues: https://github.com/flycheck/flycheck/pull/1651, https://github.com/flycheck/flycheck/issues/1722, https://github.com/flycheck/flycheck/issues/1711

We had productive conversations about related problems in flake8 with @vikigenius in the past; @vikigenius any idea about this one?

**Comment by cpitclaudel** [comment]
Thanks a lot @pedroqueiroga for reporting this, btw.  Question: do you typically use a setup.cfg or something similar at the root of your project?

**Comment by pedroqueiroga** [comment]
Nice!

> Thanks a lot @pedroqueiroga for reporting this, btw. Question: do you typically use a setup.cfg or something similar at the root of your project?

Thanks too. I have just started using setup.cfg files, I haven't been using it for long, but think I'll stick to it. Could the algorithm for the root folder be an ascent to the first directory there is no __init__.py in? It would be much the same as the directory with setup.cfg, I guess, but would only require using __init__.py, which pylint already requires if using the form `pylint pkg/foo.py`, instead of requiring a setup.cfg.

If the folder structure is something like this:
```
dir
├── weird_folder
│   └── entry.py
└── pkg
    ├── __init__.py
    ├── foo.py
    └── bar.py
```

then the root directory would be thought to be `dir`, and using `python weird_folder/entry.py` fails in much the same way as `pylint weird_folder/entry.py`.

I'm not sure if there exists something like `locate-absence-of-dominating-file`

**Comment by cpitclaudel** [comment]
> Could the algorithm for the root folder be an ascent to the first directory there is no __init__.py in? 

It could, but if that's reasonable to do… then why doesn't pylint do it?
We are running into a pretty fundamental problem here, one that's not directly related to Flycheck — it would be good to understand what the pylint folks expect us to do, but that would require reading the docs more closely before movign to their bug tracker, because we might just have missed something in the docs.

**Comment by pedroqueiroga** [comment]
Hey I think I found something that might be helpful:
[epylint](https://github.com/PyCQA/pylint/blob/master/pylint/epylint.py#L27)

It talks about exactly what we are going through right now:
> Emacs and Flymake compatible Pylint.
This script is for integration with emacs and is compatible with flymake mode.
epylint walks out of python packages before invoking pylint. This avoids
reporting import errors that occur when a module within a package uses the
absolute import path to get another module within this package.
For example:
    - Suppose a package is structured as
        a/__init__.py
        a/b/x.py
        a/c/y.py
   \- Then if y.py imports x as "from a.b import x" the following produces pylint
     errors
       cd a/c; pylint y.py
   \- The following obviously doesn't
       pylint a/c/y.py
   \- As this script will be invoked by emacs within the directory of the file
     we are checking we need to go out of it to avoid these false positives.
You may also use py_run to run pylint with desired options and get back (or not)
its output.

**Comment by pedroqueiroga** [comment]
```
(testvenv) pedro@torradeira-de-pedro:pkg$ python -c 'from pylint.epylint import py_run; py_run("bar.py")'
************* Module pkg.bar
 bar.py:1: convention (C0102, blacklisted-name, ) Black listed name "bar"
 bar.py:1: warning (W0611, unused-import, ) Unused hello_world imported from foo
 
 ----------------------------------------------------------------------
 Your code has been rated at -10.00/10 (previous run: -10.00/10, +0.00)
 
```

Here is a small proof of concept, I am not being able to extract the report from epylint's CLI

**Comment by pedroqueiroga** [comment]
very slowly I managed to achieve something similar to what is being done (I think) in flycheck:

```
python -c 'import sys;sys.path.pop(0);import runpy;runpy.run_module("pylint.epylint")["Run"]()' bar.py
************* Module pkg.bar
 bar.py:1: convention (C0102, blacklisted-name, ) Black listed name "bar"
 bar.py:1: warning (W0611, unused-import, ) Unused hello_world imported from foo
 
 ----------------------------------------------------------------------
 Your code has been rated at -10.00/10 (previous run: -10.00/10, +0.00)
 ```

**Comment by cpitclaudel** [comment]
Wow, great sleuthing work!  I'm looping in @thisch as well, who may be able to provide valuable input as well. (Hi @thisch! we're discussing a pylint bug where Flycheck is calling pylint from the wrong directory, and get spurious issues; turns out that epylint has a fix for that, so we're pondering the next move)

It would definitely be acceptable to move to epylint.  But! before we do this, I think it would be good to ask for guidance from the pylint folks (and we've done our homework now, so I think that's acceptable — thanks a lot!).  Here are the questions I think we need help on:

* What is the "right way" to run pylint on your concrete example?
* Is the algorithm implemented by epylint (which is essentially going up until there are no more `__init__.py` files) robust?  If so, why isn't it used in pylint itself by default?
* Is epylint supported, and will it remain so in the future?  Does it support all the same options as pylint, including support for reading files from stdin? (see https://github.com/PyCQA/pylint/pull/2746 and https://github.com/flycheck/flycheck/pull/1546 — I worry about this because file names are handled differently there) 
* What should Flycheck do in general: call epylint directly?  Or call pylint directly from a working directory calculated as done by epylint? Or something else?  We've been unsure for a while, esp. regarding which directory to run from, so it would be great to get a definitive answer.  Additionally, I'm assuming that other editors run into the same issue, but the tool being called epylint suggests that there's something special about Emacs ^^

@pedroqueiroga would you like to open an issue on the pylint tracker to get help with that? Ideally with a link back here.

**Comment by pedroqueiroga** [comment]
I tried a naive workaround, which I'll place below. It seems to work, but it is ugly. 
```lisp
(flycheck-define-checker python-pylint
  "A Python syntax and style checker using Pylint.

This syntax checker requires Pylint 1.0 or newer.

See URL `https://www.pylint.org/'."
  ;; --reports=n disables the scoring report.
  ;; Not calling pylint directly makes it easier to switch between different
  ;; Python versions; see https://github.com/flycheck/flycheck/issues/1055.
  :command ("python3"
            (eval (let ((result (flycheck-python-module-args
                                 'python-pylint "pylint.epylint")))
                    (list (first result)
                          (concat (second result) "[\"Run\"]()"))))
            ;; Need `source-inplace' for relative imports (e.g. `from .foo
            ;; import bar'), see https://github.com/flycheck/flycheck/issues/280
            source-inplace
            "--reports=n"
            "--output-format=json"
            (config-file "--rcfile=" flycheck-pylintrc concat))
  :error-parser flycheck-parse-pylint
  :enabled (lambda ()
             (or (not (flycheck-python-needs-module-p 'python-pylint))
                 (flycheck-python-find-module 'python-pylint "pylint")))
  :verify (lambda (_) (flycheck-python-verify-module 'python-pylint "pylint"))
  :modes python-mode
  :next-checkers ((warning . python-mypy)))
```

**Comment by pedroqueiroga** [comment]
> @pedroqueiroga would you like to open an issue on the pylint tracker to get help with that? Ideally with a link back here.

Ok!

**Comment by cpitclaudel** [comment]
Thanks a lot!  I think one of the issue links is wrong (pointing to an issue in pylint instead of here)

**Comment by cpitclaudel** [comment]
Btw @pedroqueiroga , is this behavior new in the latest pylint? I've started seeing it after upgrading, too.

**Comment by cpitclaudel** [comment]
Hi @pedroqueiroga , thanks a lot for your patience.  Since we haven't heard from the pylint folks, I've implemented the epylint directory resolution logic on our side.  Do you think you could try the branch at https://github.com/flycheck/flycheck/tree/cpitclaudel_pylint-working-directory ?

**Comment by FabriceSalvaire** [comment]
Sorry in advance ... I have not read the whole discussion ... 

but I remember that in the past, I had a wrapper hack to find the root of the source ... (epylint folk)

Pylint load and compile a Python file like Python do, it just does not run it

Thus if we have the module hiearchy  _Foo/Bar/Toto.py_  we must run pylint in the parent directory of _Foo_ and not in _Toto_ else Python will not resolve the module.

But it can be more tricky if we have several packages ... We must run pylint at the top source directory, else Python can pickup an installed file or nothing.  I always set the Python path before to open Emacs.  And I am pretty sure that IDE like PyCharm define a root directory.

**Comment by pedroqueiroga** [comment]
Pretty neat! Sorry for taking so long to respond. Works like a charm here on my end! Thanks a lot, @cpitclaudel

**Comment by cpitclaudel** [comment]
Thanks a lot for taking the time to report the issue and for your help brainstorming, testing, and debugging.

**Comment by leotrs** [comment]
From skimming this discussion it sounds like a fix was found. Any info on when this is being shipped? Thanks!

**Comment by bkhl** [comment]
@cpitclaudel I also tried that patch but it didn't work for me. I've got plenty of projects where there are several levels of subdirectories, some modules and some not.

For now I'm using a fork over here where I've used the same approach but instead look for `pyproject.toml` or `setup.py` to locate the root.

https://github.com/bkhl/flycheck/commit/5164c765b7e35b3c6b65eeb9ab899fb0377305ea

Possibly this should be made an explicitly "overridable" function? I imagine other people might prefer to hook in a function from projectile or something like that.

**Comment by g-braeunlich** [comment]
Hey guys. Still having the issue.
I am able to work around the issue by using `pylint.epylint` (as @pedroqueiroga suggested).
(If I just go to `flycheck.el`, and in `(flycheck-define-checker python-pylint`, change
```
(eval (flycheck-python-module-args 'python-pylint "pylint"))
```
to
```
(eval (flycheck-python-module-args 'python-pylint "pylint.epylint"))
```

Therefore: would not it be a good ad hoc solution to make the module configurable?

**Comment by g-braeunlich** [comment]
Actually, what is `(eval (flycheck-python-module-args 'python-pylint "pylint"))` for?
In my case, it generates
```
-c 'import sys;sys.path.pop(0);import runpy;runpy.run_module("pylint.epylint")'
```
This does not seem to work.
Replacing this with
```
"-m" "pylint.epylint"
```
does the job

**Comment by g-braeunlich** [comment]
Also make sure, that the first argument passed to `epylint` is the file name (this is different from the usual `pylint` command)
